### PR TITLE
WM-1764, WM-1862: Ad hoc WDS URL and authorized requests

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

**Review by mspector — APPROVED** [review]
Based on your suggestions in your walkthrough, I focused on `findUrlForWds` and the associated tests. It looks good to me, based on what I know about the algorithm.

The only thing that stuck out to me was the choice to focus the tests on `findUrlForWds` as opposed to the inner function `findBestAppForWds`. Just curious if that was a conscious choice, or if `findUrlForWds` just felt like the right minimum testable unit

File: service/src/main/java/bio/terra/cbas/dependencies/leonardo/AppUtils.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+  public AppUtils(
+      LeonardoServerConfiguration leonardoServerConfiguration,
+      WdsServerConfiguration wdsServerConfiguration) {
+    this.leonardoServerConfiguration = leonardoServerConfiguration;
+    this.wdsServerConfiguration = wdsServerConfiguration;
+  }
+
+  int appComparisonFunction(ListAppResponse a, ListAppResponse b) {
+    // Name scores: correct name always wins
```
Comment (salonishah11) [inline service/src/main/java/bio/terra/cbas/dependencies/leonardo/AppUtils.java:30]: It seems [in Terra UI app type along with app name check](https://github.com/DataBiosphere/terra-ui/blob/09fc15b510521e7443c2ab9fda2bf994b78bb27d/src/libs/ajax/data-table-providers/WdsDataTableProvider.ts#L107-L112) takes precedence over other methods to find the WDS proxy url. Maybe we should also include that here so that `WDS` app type in given higher precedence?

Comment (cjllanwarne) [reply service/src/main/java/bio/terra/cbas/dependencies/leonardo/AppUtils.java]: 👍

**Comment by cjllanwarne** [comment]
@mspector re 
> "The only thing that stuck out to me was the choice to focus the tests on findUrlForWds as opposed to the inner function findBestAppForWds. Just curious if that was a conscious choice, or if findUrlForWds just felt like the right minimum testable unit"

Basically, yes. I'm treating this `findUrlForWds` as a function from "list of apps" to "single URL" and testing it as a unit. I think that's (a) acceptable because it hasn't got any external dependencies and could be thought of as a pure mathematical function, and (b) appropriate because it tests the interface to this class, which is unlikely to ever change, even if the internal function layout moves around. But there is a risk that I'm not testing some edge case because of all the setup involved, so let me know if you think there's some coverages missing that would be easy enough to test at a lower level?

File: service/src/main/java/bio/terra/cbas/dependencies/common/ScheduledHealthChecker.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      CromwellService cromwellService, WdsService wdsService, LeonardoService leonardoService) {
+    this.healthCheckStatuses = new ConcurrentHashMap<>();
+    this.healthCheckSystems =
+        Map.of(
+            "cromwell", cromwellService,
+            "wds", wdsService,
+            "leonardo", leonardoService);
+  }
+
+  @Scheduled(fixedDelay = 5 * 60 * 1000)
```
Comment (salonishah11) [inline service/src/main/java/bio/terra/cbas/dependencies/common/ScheduledHealthChecker.java:38]: Should this be made configurable?

**Comment by mspector** [comment]
> so let me know if you think there's some coverages missing that would be easy enough to test at a lower level?

No need as far as I'm concerned. Thanks for explaining!

File: service/src/main/java/bio/terra/cbas/dependencies/leonardo/AppUtils.java [hunk 3]
```diff
@@ -29,7 +29,16 @@ public AppUtils(
   }
 
   int appComparisonFunction(ListAppResponse a, ListAppResponse b) {
-    // Name scores: correct name always wins
+
+    // First criteria: Prefer apps with the WDS app type.
+    // NB: Negative because lower index is better
```
Comment (mspector) [inline service/src/main/java/bio/terra/cbas/dependencies/leonardo/AppUtils.java:34]: 👍

File: service/src/test/java/bio/terra/cbas/dependencies/leonardo/TestAppUtils.java [hunk 4]
```diff
@@ -110,6 +109,14 @@ void throwIfBestAppHasNoWDS() {
     assertThrows(DependencyNotAvailableException.class, () -> au.findUrlForWds(apps));
   }
 
+  private void permuteAndTest(List<ListAppResponse> apps, String expectedUrl) throws Exception {
```
Comment (mspector) [inline service/src/test/java/bio/terra/cbas/dependencies/leonardo/TestAppUtils.java:109]: I like this
